aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,670 @@
+    struct VeiledCoinStore<phantom CoinType> has key {
+        /// A ElGamal ciphertext of a value v \in [0, 2^{32}), an invariant that is enforced throughout the code.
+        private_balance: CompressedCiphertext,
+        deposit_events: EventHandle<DepositEvent>,
+        withdraw_events: EventHandle<WithdrawEvent>,
+	pubkey: elgamal::Pubkey,
+    }
+
+    /// Holds a signer capability for the resource account created when initializing this module. This account houses a
+    /// coin::CoinStore<T> for every type of coin T that is veiled.
+    struct VeiledCoinMinter has store, key {
+        signer_cap: account::SignerCapability,
+    }
+
+    /// Event emitted when some amount of a coin is deposited into an account.
+    struct DepositEvent has drop, store {
+        // We cannot leak any information about how much has been deposited.
+    }
+
+    /// Event emitted when some amount of a coin is withdrawn from an account.
+    struct WithdrawEvent has drop, store {
+        // We cannot leak any information about how much has been withdrawn.
+    }
+
+    //
+    // Module initialization, done only once when this module is first published on the blockchain
+    //
+
+    /// Initializes a so-called "resource" account which will maintain a coin::CoinStore<T> resource for all Coin<T>'s
+    /// that have been converted into a VeiledCoin<T>.
+    fun init_module(deployer: &signer) {
+        assert!(bulletproofs::get_max_range_bits() >= MAX_BITS_IN_VALUE, ERANGE_PROOF_SYSTEM_HAS_INSUFFICIENT_RANGE);
+
+        // Create the resource account. This will allow this module to later obtain a `signer` for this account and
+        // transfer Coin<T>'s into its CoinStore<T> before minting a VeiledCoin<T>.
+        let (_resource, signer_cap) = account::create_resource_account(deployer, vector::empty());
+
+        move_to(deployer,
+            VeiledCoinMinter {
+                signer_cap
+            }
+        )
+    }
+
+    //
+    // Entry functions
+    //
+
+    /// Takes `amount` of `coin::Coin<CoinType>` coins from `sender`, wraps them inside a `VeiledCoin<CoinType>` and
+    /// sends them back to `sender`.
+    public entry fun wrap<CoinType>(sender: &signer, amount: u64) acquires VeiledCoinMinter, VeiledCoinStore {
+        wrap_to<CoinType>(sender, signer::address_of(sender), amount)
+    }
+
+    /// Takes `amount` of `coin::Coin<CoinType>` coins from `sender`, wraps them inside a `VeiledCoin<CoinType>` and\
+    /// sends to `recipient`.
+    public entry fun wrap_to<CoinType>(sender: &signer, recipient: address, amount: u64) acquires VeiledCoinMinter, VeiledCoinStore {
+        let c = coin::withdraw<CoinType>(sender, amount);
+
+        let vc = wrap_from_coin(c);
+
+        deposit<CoinType>(recipient, vc)
+    }
+
+    /// Takes `amount` of `VeiledCoin<CoinType>` coins from `sender`, unwraps them to a coin::Coin<CoinType>,
+    /// and sends them back to `sender`.`amount_ct` must be an el-gamal ciphertext encrypting `amount`
```

Comment:
Typo: el-gamal --> ElGamal

Reply:
What do you mean by "It is not possible for this function to be made private."?

Reply:
Rephrase "It is not possible for this function to be made private" as "Note that this function inherently leaks the `amount`, since privacy of the balance is maintained.

Reply:
Changed to "Note that this function inherently leaks `amount`. Privacy of the remaining balance is maintained."